src/StateAwareTrait.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+namespace Dhii\State;
+
+use Dhii\Util\String\StringableInterface as Stringable;
+use Exception as RootException;
+use InvalidArgumentException;
+
+/**
+ * Common functionality for objects that are aware of a state.
+ *
+ * @since [*next-version*]
+ */
+trait StateAwareTrait
+{
+    /**
+     * The state.
+     *
+     * @since [*next-version*]
+     *
+     * @var string|Stringable|null
+     */
+    protected $state;
+
+    /**
+     * Retrieves the state associated with this instance.
+     *
+     * @since [*next-version*]
+     *
+     * @return string|Stringable|null The state, if any.
+     */
+    protected function _getState()
+    {
+        return $this->state;
+    }
+
+    /**
+     * Sets the state for this instance.
+     *
+     * @since [*next-version*]
+     *
+     * @param string|Stringable|null $state The state, or null.
+     */
+    protected function _setState($state)
+    {
+        if ($state !== null && !is_string($state) && !($state instanceof Stringable)) {
+            throw $this->_createInvalidArgumentException(
+                $this->__('Argument is not a valid state or null.'),
```

Comment:
IMHO the "or null" is extra here, because `null` is a valid state. Also, I would say it's best to omit the trailing fullstop.